### DISCUSSION TITLE
2D FFT of laser pulse

### DISCUSSION THREAD

**Discussion by laxmisubramanyam** [SmileiPIC/Smilei#764] [Q&A] [answered]
Hi Friends,

Please, is it possible to help me to understand how to do 2D FFT of a laser pulse. For, example, the pulse is defined on the spatial grid as Ey(x,y). I can do 1D FFT (by taking line out of the electric field) but I messed up to extend that approach to 2 dimension.

I want to do spectral analysis similar to the one performed in Fig 4 of the attached article.

Any help will be very appreciable.

Thanks a lot.
Laxmi

[POP_Giulietti_2013.pdf](https://github.com/user-attachments/files/17709391/POP_Giulietti_2013.pdf)

**Comment by mccoys** [comment]
I moved this to discussions as it is not an issue with smilei.

**Reply by laxmisubramanyam** [reply]
Hi, 

Can I get some help on 2D FFT ...

Thank you.

**Comment by xxirii** [comment]
Hi,

Python or Matlab can perform 2D (even more) FFTs in few lines. Have you tried these technics?

**Comment by laxmisubramanyam** [comment]
Hi @xxirii,

I tried as show below. But in both cases, I was not able to see as done in 4 Fig. of the above paper. My main aim is to see which part of the laser pulse is undergoing frequency change (front, middle or back)

I am giving here both 1-D spectrum calculation and 2-D spectrum calculation (both in Matlab)

```First: 1D case```

```
field_dist = Ex2D;   % suppose Ex2D is the 2D electric field distribution of the laser pulse is this

x_max = length(x_axis);    % length of x-coordinates
y_max = length(y_axis);    %length of y-coordinates

line_out_Efield_y = sum(field_dist,2);   % Instead of summing up along all the rows corresponding to every transverse coordinate, one can even just take a line out along "y = 0"

% General prepartion for spectrum calculation

c = 3.0e8;
wl = 0.8e-6;				% wavelength
fl = c/wl;				        % frequency
kl = 2*pi/wl;			       % wave-vector

Nfft = length(x1d);			% window length (number of samples)

dx=(x1d(2)-x1d(1));			

Ts = dx/c;				    % time increment per sample

Fs = 1/Ts;				    % sampling frequency ... sample per unit time

f = (0:Nfft-1)*Fs/Nfft;			    % frequency range for data

%f = (-Nfft/2:Nfft/2-1)*Fs/Nfft;	% frequency range for data

lamda = c./f; 

Efield1d_y_spectrum = abs(fft(line_out_Efield_y,Nfft));  % Spectrum of the laser pulse

% plotting set up

linewidth = 1;
fontsize = 8;

subplot(2,1,1)

plot(x1d*1e6,line_out_Efield_y);
xlabel('x ( \mum )','FontSize',fontsize); ylabel('E-field (V/m)','FontSize',fontsize);


subplot(2,1,2)

plot(1e9*lamda, Efield1d_y_spectrum)
set(gca,'XLim',[200 1400]);
xlabel('Wavelength (nm)','FontSize',fontsize); ylabel('Intensity (a.u.)','FontSize',fontsize);
```

```Secondt: 2D case```

```
field_dist = Ex2D;

c = 3.0e8;
wl = 10.0e-6;				% wavelength of the laser pulse
fl = c/wl;				% frequency of the laser pulse
kl = 2.0*pi/wl;				% wave-vector of the laser pulse
omega_l = 2.0*pi*fl;

%Nfft_x = length(x1d);

Nfft_x = x_max;

dx = (x1d(2)-x1d(1));			% fft set up along x-direction			

Ts_x = dx/c;

Fs_x = 1/Ts_x;

%f_x = (1:Nfft_x)*Fs_x/Nfft_x*Nfft_x/dx;

f_x = (-Nfft_x/2:Nfft_x/2-1)*Fs_x / (Nfft_x);

lamda_x = c./f_x;

k_x = (2*pi)./lamda_x;


%Nfft_y = length(y1d);			% fft set up along y-direction

Nfft_y = y_max;

dy=(y1d(2)-y1d(1));			

Ts_y = dy/c;

Fs_y = 1/Ts_y;
    
%f_y = (1:Nfft_y)*Fs_y/Nfft_y;

f_y = (-Nfft_y/2:Nfft_y/2-1)*Fs_y / (Nfft_y);

lamda_y = c./f_y; 

k_y = (2*pi)./lamda_y;

Y_freq = Fs_y/Nfft_y * (-Nfft_y/2:Nfft_y/2-1);
X_freq = Fs_x/Nfft_x * (-Nfft_x/2:Nfft_x/2-1);
                        
Efield_fft = abs(fftshift(fft2(field_dist,Nfft_y,Nfft_x)));

linewidth = 1;
fontsize = 8;

length(k_x*wl)
length(yaxs)

colormap(flipud(hot))

imagesc(k_x*wl, k_y*wl, Efield_fft)
%imagesc(k_x*wl, k_y*wl, log2(Efield_fft))

```

**Comment by xxirii** [comment]
Hello, 

I have looked quickly at the paper. I think this is not 2D fft but 1D fft performed for each time-space position $X$. This is not clearly explained in the paper but the sentence `Spectra of Ey field on axis. For each Xi , the spectrum of the field
times a super-Gaussian function, centered on $X_i$ , is given.` suggests that they take the laser Ey field along the x-axis for a given y transverse value I think. Then they use a super-Gaussian window to scan the wave along the x-axis and create the space-time dependency. They perform a 1D fft for each position $X_i$ corresponding of the fft of the multiplication of the Gaussian profile with the 1D Ey profile. Nonetheless, even if this is a classical method, I admin that it is not clearly explained. It may be good to directly ask the authors to be sure.

They do not specify the properties of the super-Gaussian (which order, fwhm, etc) so you will miss some parameters to reproduce the results. I strongly recommend to contact the authors. Being able to reproduce a paper result is very important so they should give you more details.

**Comment by laxmisubramanyam** [comment]
Hi @xxirii,

Thanks a lot for your time and suggestion. I will try to contact them but I already know from the first author that he didn't done that simulation. 

Do you have any idea how one can do such presentation of frequency change ?

**Comment by Z10Frank** [comment, marked as answer]
I agree with @xxirii . 
This could be a Wigner transform of the field.

**Reply by laxmisubramanyam** [reply]
@Z10Frank, 

Is there any example which I can follow and try to do Wigner transform. Could you please help me, if possible.

**Reply by Z10Frank** [reply]
I have never implemented or used this transform, so I don't have an example in mind.
My suggestion is to either:

- use the definition of Wigner transform and implement it in Python or your language of choice "from scratch": https://en.wikipedia.org/wiki/Wigner_distribution_function In some ways it is similar to a Fourier transform.

Or

- find a library online that performs the Wigner transform in Python or your language of choice.

**Reply by laxmisubramanyam** [reply]
@Z10Frank,

Thank you. This probably will help.

**Comment by laxmisubramanyam** [comment]
@Z10Frank , eventually I was able to do it. Thanks for the tip !!

Best,
Laxmi
